### PR TITLE
Python editor docking and undocking

### PR DESCRIPTION
**Pull request by reddyashish**
### Purpose

Python editor docking and undocking. https://jira.autodesk.com/browse/DYN-3393

DockWindowInSideBar() is introduced to dock a window to the right side bar.
UndockWindow() is to undock a tab from the right side bar and call the necessary action to launch the original window.

python docking new

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated
- [ ] This PR contains no files larger than 50 MB 

### Release Notes
Python editor docking and undocking

### Reviewers
@QilongTang @zeusongit

### CODE REVIEW

**Comment by QilongTang** [comment]
I feel like we need a slightly larger dock icon on Python editor? @Jingyi-Wen

**Comment by Jingyi-Wen** [comment]
> I feel like we need a slightly larger dock icon on Python editor? @Jingyi-Wen

Hmm, this does look weird. Let me make a bigger one. 
Also, I'd suggest putting this icon in between the minimize and maximize.

**Comment by reddyashish** [comment]
Updated the changes with new icon and fixed editor functionality when it is docked. Working on tests now.

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs [hunk 1]
```diff
@@ -604,7 +604,18 @@ private void CloseButton_OnClick(object sender, RoutedEventArgs e)
         // Handles Close button on the Warning bar
         private void CloseWarningBarButton_OnClick(object sender, RoutedEventArgs e)
         {
-            this.Close();
+            var senderContext = (sender as Button).DataContext;
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:607]: null check?

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs [hunk 2]
```diff
@@ -567,7 +604,18 @@ private void CloseButton_OnClick(object sender, RoutedEventArgs e)
...
+
+            if (senderContext.Equals(this))
+            {
+                Close();
+            }
+            else // Close the right side extension tab if the close button is clicked on the docked editor. 
+            {
+                var dynamoView = Owner as DynamoView;
+                TabItem tabItem = dynamoView.ExtensionTabItems.OfType<TabItem>().SingleOrDefault(n => n.Uid.ToString() == NodeModel.GUID.ToString());
+                dynamoView.CloseExtensionTab(tabItem);
```
Comment (QilongTang) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:617]: Why do we close the whole tab as well?

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs [hunk 3]
```diff
@@ -461,18 +492,24 @@ private void OnEngineChanged(object sender, System.Windows.Controls.SelectionCha
...
+            var dybnamoView = Owner as DynamoView;
 
-            Analytics.TrackEvent(
-                Dynamo.Logging.Actions.Close,
-                Dynamo.Logging.Categories.PythonOperations);
-            
-            editText.TextChanged -= EditTextOnTextChanged;
-            FoldingManager.Uninstall(foldingManager);
-            foldingManager = null;
+            if (!dynamoViewModel.DockedWindows.ContainsKey(Uid))
```
Comment (QilongTang) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:497]: Would you add some comments?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 4]
```diff
@@ -464,6 +465,66 @@ private TabItem AddExtensionTab(IViewExtension viewExtension, UIElement content)
...
+            tab.Content = content;
+
+            //Insert the tab at the end
+            ExtensionTabItems.Insert(ExtensionTabItems.Count, tab);
+
+            tabDynamic.DataContext = ExtensionTabItems;
+            tabDynamic.SelectedItem = tab;
+
+            Analytics.TrackEvent(Actions.Dock,
+                     Categories.ViewExtensionOperations, tab.Header.ToString());
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:517]: You may want to track it differently here or not track it for now

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 5]
```diff
@@ -491,22 +552,40 @@ internal void CloseExtensionControl(IViewExtension viewExtension)
...
-            TabItem tabitem = ExtensionTabItems.OfType<TabItem>().SingleOrDefault(n => n.Header.ToString() == tabName);
+            string tabId = (sender as Button).Uid.ToString();
+            TabItem tabitem = ExtensionTabItems.OfType<TabItem>().SingleOrDefault(n => n.Uid.ToString() == tabId);
 
             if (tabitem.Tag is ViewExtensionBase viewExtensionBase)
             {
                 viewExtensionBase.Closed();
             }
 
+            dynamoViewModel.DockedWindows.TryGetValue(tabitem.Uid, out NodeModel nodeModel);
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:563]: Instead of grabbing the node model, it seems we can simply search the node using the uid?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 6]
```diff
@@ -71,6 +70,17 @@ public partial class DynamoViewModel : ViewModelBase, IDynamoViewModel
...
+
+        /// <summary>
+        /// Docked windows in sidebar
+        /// </summary>
+        internal Dictionary<string, NodeModel> DockedWindows { get; set; } = new Dictionary<string, NodeModel>();
+
+        /// <summary>
+        /// Docked node windows in sidebar
+        /// </summary>
+        internal Dictionary<string, ViewExtensionDisplayMode> NodeWindowStates { get; set; } = new Dictionary<string, ViewExtensionDisplayMode>();
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs:82]: To me, looks like we only need the second dictionary here? It worries me if we store the NodeModel references in `DockedWindows` because if the Python node is removed in canvas, we still need to handle the disposal in addition?

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs (hunk 2 above)
Comment (reddyashish) [reply src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:617]: There is a close button on the warning message and that would close the tab without saving. Like how the script editor window is closed.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 5 above)
Comment (reddyashish) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:563]: yeah good point, search the nodes in the workspace with the ID?

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs [hunk 7]
```diff
@@ -464,6 +465,66 @@ private TabItem AddExtensionTab(IViewExtension viewExtension, UIElement content)
...
+            var content = window.Content;
+
+            // creates a new tab item
+            var tab = new TabItem();
+            tab.Header = window.Title;
+            tab.Tag = window;
+            tab.Uid = window.Uid;
+            tab.HeaderTemplate = tabDynamic.FindResource("TabHeader") as DataTemplate;
+
+            // setting the extension UI to the current tab content
```
[the remark this answers is not in the document]
Comment (reddyashish) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:500]: Done

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 4 above)
Comment (reddyashish) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs:517]: Yeah removed it for now.

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs (hunk 3 above)
Comment (reddyashish) [reply src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:497]: Added comment:
"When the script editor is docked, we don't to dispose the editor functionality.
Dispose it only when the window is closed."

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs (hunk 1 above)
Comment (reddyashish) [reply src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml.cs:607]: Done.

File: src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs (hunk 5 above)
Comment (reddyashish) [reply src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs]: Using just a HashSet now to get the current docked windows. Not storing the nodemodel object now and getting that info from workspace nodes.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 6 above)
Comment (reddyashish) [reply src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Yes addressed this in the above comment.

**Comment by reddyashish** [comment]
@QilongTang The tab name is based on the node name but it won't rename the tab it is docked. When it is opened again, it will have the new node name. Do you expect it to rename when it is docked?

**Comment by QilongTang** [comment]
> @QilongTang The tab name is based on the node name but it won't rename the tab if it is docked. When the tab is opened again, it will have the new node name. Do you expect it to rename when it is docked?

This probably can be follow up task. But I expect the name in sync

**Comment by reddyashish** [comment]
one failing test. Looking into it.

**Comment by reddyashish** [comment]
Fixed an edge case, edit context menu on the script now works fine when docked.

**Review by QilongTang — APPROVED** [review]
